In [4]:
import pandas as pd

In [5]:
filename = "./datasets/default of credit card clients.csv"
dataset = pd.read_csv(filename, delimiter=";")


In [6]:
dataset

X1  X2  X3  X4  X5  X6  X7  X8  X9  X10  ...    X15    X16    X17  \
0       20000   2   2   1  24   2   2  -1  -1   -2  ...      0      0      0   
1      120000   2   2   2  26  -1   2   0   0    0  ...   3272   3455   3261   
2       90000   2   2   2  34   0   0   0   0    0  ...  14331  14948  15549   
3       50000   2   2   1  37   0   0   0   0    0  ...  28314  28959  29547   
4       50000   1   2   1  57  -1   0  -1   0    0  ...  20940  19146  19131   
...       ...  ..  ..  ..  ..  ..  ..  ..  ..  ...  ...    ...    ...    ...   
29995  220000   1   3   1  39   0   0   0   0    0  ...  88004  31237  15980   
29996  150000   1   3   2  43  -1  -1  -1  -1    0  ...   8979   5190      0   
29997   30000   1   2   2  37   4   3   2  -1    0  ...  20878  20582  19357   
29998   80000   1   3   1  41   1  -1   0   0    0  ...  52774  11855  48944   
29999   50000   1   2   1  46   0   0   0   0    0  ...  36535  32428  15313   

         X18    X19    X20   X21    X22   X23  Y  
0          0    689      0     0      0     0  1  
1          0   1000   1000  1000      0  2000  1  
2       1518   1500   1000  1000   1000  5000  0  
3       2000   2019   1200  1100   1069  1000  0  
4       2000  36681  10000  9000    689   679  0  
...      ...    ...    ...   ...    ...   ... ..  
29995   8500  20000   5003  3047   5000  1000  0  
29996   1837   3526   8998   129      0     0  0  
29997      0      0  22000  4200   2000  3100  1  
29998  85900   3409   1178  1926  52964  1804  1  
29999   2078   1800   1430  1000   1000  1000  1  

[30000 rows x 24 columns]

# Analysis of values and corresponding feature intervals

In [9]:
print("Given credit (NT$)\n", dataset["X1"].value_counts().sort_index())
print("Gender\n", dataset["X2"].value_counts().sort_index())
print("Education\n", dataset["X3"].value_counts().sort_index())
print("Marital status\n", dataset["X4"].value_counts().sort_index())
print("Age\n", dataset["X5"].value_counts().sort_index())
print("Past, monthly bill (-1)\n", dataset["X12"].value_counts().sort_index())
print("Past, monthly bill (-2)\n", dataset["X13"].value_counts().sort_index())
print("Past, monthly bill (-3)\n", dataset["X14"].value_counts().sort_index())
print("Past, monthly bill (-4)\n", dataset["X15"].value_counts().sort_index())
print("Past, monthly bill (-5)\n", dataset["X16"].value_counts().sort_index())
print("Past, monthly bill (-6)\n", dataset["X17"].value_counts().sort_index())
print("Prev. payment in NT$ (-1)\n", dataset["X18"].value_counts().sort_index())
print("Prev. payment in NT$ (-2)\n", dataset["X19"].value_counts().sort_index())
print("Prev. payment in NT$ (-3)\n", dataset["X20"].value_counts().sort_index())
print("Prev. payment in NT$ (-4)\n", dataset["X21"].value_counts().sort_index())
print("Prev. payment in NT$ (-5)\n", dataset["X22"].value_counts().sort_index())
print("Prev. payment in NT$ (-6)\n", dataset["X23"].value_counts().sort_index())

Given credit (NT$)
 10000       493
16000         2
20000      1976
30000      1610
40000       230
           ... 
750000        4
760000        1
780000        2
800000        2
1000000       1
Name: X1, Length: 81, dtype: int64
Gender
 1    11888
2    18112
Name: X2, dtype: int64
Education
 0       14
1    10585
2    14030
3     4917
4      123
5      280
6       51
Name: X3, dtype: int64
Marital status
 0       54
1    13659
2    15964
3      323
Name: X4, dtype: int64
Age
 21      67
22     560
23     931
24    1127
25    1186
26    1256
27    1477
28    1409
29    1605
30    1395
31    1217
32    1158
33    1146
34    1162
35    1113
36    1108
37    1041
38     944
39     954
40     870
41     824
42     794
43     670
44     700
45     617
46     570
47     501
48     466
49     452
50     411
51     340
52     304
53     325
54     247
55     209
56     178
57     122
58     122
59      83
60      67
61      56
62      44
63      31
64      31
65      24
66      25
67      16


## Resulting analysis
X3 has values not defined in the dictionary.
1. 0 - 14  values
2. 5 - 280 values
3. 6 - 51 values

X4 is in the same situation
1. 0 - 54 values

The best course of action for these values would be to remove them, since they are only a few cases and it is not defined what they mean.

Features X12 to X17 have records with negative values. These features are related to the bill statement

In [11]:
dataset["Y"].value_counts().sort_index()


0    23364
1     6636
Name: Y, dtype: int64

## Comparing to the described values in the dictionary

The interval of possible values of the features X6-X11 is [-1, 9] (excluding 0).

The analysis below indicates a big discrepancy in this rule, with the majority of values being 0, a considerable amount with the value -2 and none with the value 9. This might indicate an error either in data-input or the creation of the dictionary.
Deciding based on logical intuition of the absence of a possible value across these features, along with the presence of values that should not exist, we conclude that the best course of action to take here is to:

1. Agglomerate features with values 0 into the ones with value -1
2. Increment by one these features - thus making rows with these features' values 9 possible


Tendo em conta o ponto 1, é uma quantidade considerável de registos cujo valor é 0 para as variáveis vistas.
Já que a distribuição da amostra favorece o caso onde a falha no pagamento é 0 (ou seja, o pagamento é previsto como realizado) E tendo em consideração o facto das análises realizadas até agora considerarem o PAY_1 (atraso no pagamento no mês anterior)


In [10]:
# Verification of values not described in the dictionary
print("Past, monthly payment (-1)\n", dataset["X6"].value_counts().sort_index())
print("Past, monthly payment (-2)\n", dataset["X7"].value_counts().sort_index())
print("Past, monthly payment (-3)\n", dataset["X8"].value_counts().sort_index())
print("Past, monthly payment (-4)\n", dataset["X9"].value_counts().sort_index())
print("Past, monthly payment (-5)\n", dataset["X10"].value_counts().sort_index())
print("Past, monthly payment (-6)\n", dataset["X11"].value_counts().sort_index())

Past, monthly payment (-1)
 -2     2759
-1     5686
 0    14737
 1     3688
 2     2667
 3      322
 4       76
 5       26
 6       11
 7        9
 8       19
Name: X6, dtype: int64
Past, monthly payment (-2)
 -2     3782
-1     6050
 0    15730
 1       28
 2     3927
 3      326
 4       99
 5       25
 6       12
 7       20
 8        1
Name: X7, dtype: int64
Past, monthly payment (-3)
 -2     4085
-1     5938
 0    15764
 1        4
 2     3819
 3      240
 4       76
 5       21
 6       23
 7       27
 8        3
Name: X8, dtype: int64
Past, monthly payment (-4)
 -2     4348
-1     5687
 0    16455
 1        2
 2     3159
 3      180
 4       69
 5       35
 6        5
 7       58
 8        2
Name: X9, dtype: int64
Past, monthly payment (-5)
 -2     4546
-1     5539
 0    16947
 2     2626
 3      178
 4       84
 5       17
 6        4
 7       58
 8        1
Name: X10, dtype: int64
Past, monthly payment (-6)
 -2     4895
-1     5740
 0    16286
 2     2766
 3      184
 4      